Blog posts: a deep dive into a Go compiler bug fix (#803)
mattermost/mattermost-developer-documentation: site/content/blog/2021-04-23-optimizing-go-compiler-1.md

heading: "Fixing a bug in the Go compiler as a newbie: a deep dive (I)"
description: ""
summary: "Contributing to the Go compiler is within everyone's reach. The first part of this series of posts defines the issue we are going to investigate, and introduces the parts of the Go compiler involved in its fix. We will learn about Static Single Assignment, rewrite rules and some specific incantations to hack on the Go compiler."
slug: optimizing-go-compiler-1
date: 2021-04-23T12:00:00-05:00
categories:
    - "go"
author: Alejandro García Montoro
github: agarciamontoro
community: alejandro.garcia
---

[bf48163e8f2b604f3b9e83951e331cd11edd8495](https://github.com/golang/go/commit/bf48163e8f2b604f3b9e83951e331cd11edd8495). That is the hash of one of the 47791 commits that, at the time of this writing, build the whole git history of the Go compiler. It is definitely not the most important commit there, and most of the other ones probably deserve way more attention. But I like `bf48163`. It is a good commit, I think. I may be a bit biased, of course, as it is my first contribution to Go.

Not only that, it is my first contribution _ever_ to any compiler!

> What's really interesting here is that I am not an expert in Go---nor in compilers in general---at all. Sure, I know my way through the language to write software using it, but I had never thought I could be able to modify the compiler itself.

Author: Just a sanity check: the three dashes render like this:
![image](https://user-images.githubusercontent.com/3924815/119796834-002c8e00-beda-11eb-9457-f027fb8911f6.png)

While only one dash renders as follows:
![image](https://user-images.githubusercontent.com/3924815/119796762-ee4aeb00-bed9-11eb-99ac-2e615d5d7895.png)

Is this what we want, @justinegeffen?

> What's really interesting here is that I am not an expert in Go---nor in compilers in general---at all. Sure, I know my way through the language to write software using it, but I had never thought I could be able to modify the compiler itself.

Reviewer: Great question! So they're both correct imo. I'm 0/5 as to which we use so let's go with the original. :)

The only weird line might be the first one, which is a bounds check. This forces the compiler to check that `dest[8]`, and, thus, all `dest[0]`, ..., `dest[7]`, are valid indices. We do that at the very beginning, so the compiler knows that all indices we are using in the copy are valid, for the whole scope of the function. We do not need to do such a check with `src`, because its length is known at compile time: 8 bytes.

So what's the bug? This is where things start to get interesting. The compiler is generating correct code, but it is not as optimized as it can be.

# Understanding the compiler

Before diving into the bug, let's pause for a bit to learn something more about how the Go compiler works.

Essentially, a compiler is a translator: it converts a piece of code in a language (in our case, Go) to an equivalent piece of code in another language (in our case, assembly code, whose specific instructions depend on the specific architecture we want to run the executable in). Hopefully, the meaning of what we wrote in Go (the behavior of our program), is maintained throughout the process of translation.

> In the Go compiler, this complex process is divided into phases. There are _a lot_ of them, and we will not discuss the structure in this blog post---see [`cmd/compile/README.md`](https://github.com/golang/go/blob/0c93b16d015663a60ac77900ca0dcfab92310790/src/cmd/compile/README.md) for a high-level view of the phases, and take a look directly [at the code](https://github.com/golang/go/blob/0c93b16d015663a60ac77900ca0dcfab92310790/src/cmd/compile/internal/ssa/compile.go#L426-L481) to see the complete list---. We will instead focus on one of the last phases, which converts the so-called Static Single Assignment (SSA) form into the final assembly instructions, that depend on the architecture we are compiling for.

Author: @justinegeffen, same question as before.

-   The new instruction, `(Lsh8x64 <t> n (Const64 <typ.UInt64> [log8(c)]))`, which shifts `n` to the left `log(c)` places.

It is ok if you do not understand every specific detail of the instructions, we will see them in detail a bit later in this post. Here, for example, we can note that the `8` suffixes make reference to the fact that this rule only matches multiplications of 8-bit variables. The other highlighted rules in the link to the `generic.rules` file cover the cases for variables of different bit widths.

For now, it's enough to understand that the rules match a specific pattern of instructions, and rewrite those instructions with a new one that is hopefully more efficient. Also, it is important to know that there are generic rules, as the one we just studied, which are applied regardless of the specific architecture we are compiling for, and architecture-specific rules, such as the ones in the [AMD64.rules file](https://github.com/golang/go/blob/master/src/cmd/compile/internal/ssa/gen/AMD64.rules), which take advantage of the specifics of each architecture, further optimizing the code.

## Connecting all together

> In order to investigate the issue and solve it, we will need to set up the development environment for the Go compiler. I encourage you to read the official [documentation](https://golang.org/doc/contribute): it is comprehensive and easy to understand. There are some particularities to the process, but it is really straightforward.

Reviewer: ```suggestion
In order to investigate the issue and solve it, we'll need to set up the development environment for the Go compiler. I encourage you to read the official [documentation](https://golang.org/doc/contribute): it's comprehensive and easy to understand. There are some particularities to the process, but it's generally straightforward.
```

> For the specifics we will need to use in these posts, we can summarize a couple of things. All the commands in this section assume that you have cloned the Go repository and have a shell open in its root directory.

Reviewer: ```suggestion
For the specifics we'll need to use in these posts, we can summarize a couple of things. All the commands in this section assume that you've cloned the Go repository and have a shell open in its root directory.
```

### How to apply the changes done to the rewrite rules

> We already know what the rewrite rules are, but not how they are applied. Although the rules live in a series of `.rules` files, the actual code used to match and apply them is normal Go code. With one peculiarity: it is automatically generated from the `.rules` files.

Reviewer: ```suggestion
We already know what the rewrite rules are, but not how they're applied. Although the rules live in a series of `.rules` files, the actual code used to match and apply them is normal Go code. With one peculiarity: it's automatically generated from the `.rules` files.
```

For example, the rules for AMD64 are declared in the file `src/cmd/compile/internal/ssa/gen/AMD64.rules `, but the automatically generated code that applies those rules live in `src/cmd/compile/internal/ssa/rewriteAMD64.go`.

> So: whenever we manually update a rewrite rule, we need to re-generate the corresponding Go code. Doing so it's trivial, and it comes down to this:

Reviewer: ```suggestion
So: whenever we manually update a rewrite rule, we need to re-generate the corresponding Go code. Doing so is trivial, and comes down to this:
```

```sh
cd src/cmd/compile/internal/ssa/gen/
go run *.go
```

### How to compile the compiler and test it

> And now for the fun part: how to compile the compiler. Once we have generated the automatic code, we want to build the compiler from source with the new changes. Doing it is also as trivial as calling a script:

Reviewer: ```suggestion
And now for the fun part: how to compile the compiler. Once we've generated the automatic code, we want to build the compiler from source with the new changes. Doing it is as trivial as calling a script:
```

```sh
cd src/
./make.bash
```

Once this finishes, make sure that the subsequent calls to the `go` command are using the recently created binary:

```sh
cd bin/
export PATH=`pwd`:$PATH
```

If everything was correctly setup, you should see your local Go directory when doing `which go`.

> And last, but not least, when you have applied your changes, you will want to know that you did not break anything else, so you should run the whole test suite at least once before you push your changes! Doing so is again one script call away:

Reviewer: ```suggestion
And last, but not least, when you've applied your changes, you'll want to know that you didn't break anything else, so you should run the whole test suite at least once before you push your changes! Doing so is again one script call away:
```

---
title: "Fixing a bug in the Go compiler as a newbie: a deep dive (I)"
heading: "Fixing a bug in the Go compiler as a newbie: a deep dive (I)"
description: ""
summary: "Contributing to the Go compiler is within everyone's reach. The first part of this series of posts defines the issue we are going to investigate, and introduces the parts of the Go compiler involved in its fix. We will learn about Static Single Assignment, rewrite rules and some specific incantations to hack on the Go compiler."
slug: optimizing-go-compiler-1
date: 2021-04-23T12:00:00-05:00
categories:
    - "go"
author: Alejandro García Montoro
github: agarciamontoro
community: alejandro.garcia
---

> [bf48163e8f2b604f3b9e83951e331cd11edd8495](https://github.com/golang/go/commit/bf48163e8f2b604f3b9e83951e331cd11edd8495). That is the hash of one of the 47791 commits that, at the time of this writing, build the whole git history of the Go compiler. It is definitely not the most important commit there, and most of the other ones probably deserve way more attention. But I like `bf48163`. It is a good commit, I think. I may be a bit biased, of course, as it is my first contribution to Go.

Reviewer 2: > That is the hash of one of the 47791 commits that, at the time of this writing, build the whole git history of the Go compiler.

The last part reads a bit odd to me. I understand what you are trying to say, but what does "build the whole git history" mean? Isn't it just building the Go compiler?

0/5, but perhaps something like this is equivalent:

> [bf48163e8f2b604f3b9e83951e331cd11edd8495](https://github.com/golang/go/commit/bf48163e8f2b604f3b9e83951e331cd11edd8495). That is the hash of one of the 47791 commits of the Go programming language, at the time of this writing.

Author: > The last part reads a bit odd to me. I understand what you are trying to say, but what does "build the whole git history" mean? Isn't it just building the Go compiler?

My intention with this was to say that there are 47791 commits in the git history, each and everyone contributing to create the Go compiler. It is not build as in `make build`, but in the sense of constitute, of contributing to the history of the development of the compiler.

Would it be more clear to say something like this?

> That is the hash of one of the 47791 commits that, at the time of this writing, **form/populate/build up?** the whole git history of the Go compiler.

If it's still confusing, I'll use something like what you suggested.

Reviewer 2: Right, that's more or less what I thought. It's the start of the post, so I'll leave it to whatever you feel is the best. :)

Author: Resolving this conversation too to remove some noise from the PR. Thanks again!

Not only that, it is my first contribution _ever_ to any compiler!

What's really interesting here is that I am not an expert in Go---nor in compilers in general---at all. Sure, I know my way through the language to write software using it, but I had never thought I could be able to modify the compiler itself.

And that is why I wrote this, to let you know that fixing a bug in a compiler is not only meant for a few chosen ones, but within reach of anyone.

Although this was conceived as a single post, it got a wee bit out of hand, so I decided to split it in three:

-   Part 1: what the bug looks like, and an introduction to some parts of the Go compiler.
-   [Part 2](/blog/optimizing-go-compiler-2): what the bug actually is.
-   [Part 3](/blog/optimizing-go-compiler-3): how to fix the bug.

> Do not get discouraged for the length of the content: it does not correlate with the complexity of the issue, but with the level of detail I tried to use when explaining it. I wanted to make sure I explained the ins and outs of how I approached the fix, and I do not assume any knowledge about compilers from the reader, so I tried to discuss everything in chapter and verse.

Reviewer 2: There shouldn't be a period at the end of "explaining it" I think. Maybe place it at the end of "approached the fix". But I'll leave the grammatical stuff to Justine :P

Author: Hm, not sure. For context, this was the structure in my head:

Do not get discouraged for X: it is not because of Y, but because of Z. And Z happens because I had these intentions.

But yeah, I agree that Justine has the last word!

Reviewer 2: Ah ok, I read it slightly differently.

# The issue

Ok, so let's dive into the issue itself. The bug in the compiler can be explained with the following function:

```go
import "encoding/binary"

func f(b []byte, x *[8]byte) {
	_ = b[8]
	t := binary.LittleEndian.Uint64(x[:])
	binary.LittleEndian.PutUint64(b, t)
}
```

The report said that this function compiled to a series of small `MOVx`, but that it could be optimized to use instead a couple of `MOVQ`: one for loading and one for storing.

If this sounds as gibberish, that's ok. Bear with me and we will get to what that sentence means. First, let's see what the function is supposed to do. To make things a bit easier, let's simply rewrite the variable names and add a couple of comments:

```go
import "encoding/binary"

func copyArrayToSlice(dest []byte, src *[8]byte) {
	_ = dest[8]                                // 1. bounds check
	temp := binary.LittleEndian.Uint64(src[:]) // 2. read the contents of the source array into temp
	binary.LittleEndian.PutUint64(dest, temp)  // 3. write the contents into the destiny slice
}
```

> It is now more clear what the function does. It stores the contents of the second argument, the `src` array, into the first argument, the slice `dest`. It uses the [encoding/binary](https://golang.org/pkg/encoding/binary/) package for reading and writing the 8 bytes as an unsigned integer of 64 bits. For doing the copy, we use a temporal variable, `temp`.

Reviewer 2: maybe "temporary" variable is better. "temporal" sounds like a sci-fi term :P

Author: Good catch, that's a mistake I always make!

The only weird line might be the first one, which is a bounds check. This forces the compiler to check that `dest[8]`, and, thus, all `dest[0]`, ..., `dest[7]`, are valid indices. We do that at the very beginning, so the compiler knows that all indices we are using in the copy are valid, for the whole scope of the function. We do not need to do such a check with `src`, because its length is known at compile time: 8 bytes.

So what's the bug? This is where things start to get interesting. The compiler is generating correct code, but it is not as optimized as it can be.

# Understanding the compiler

Before diving into the bug, let's pause for a bit to learn something more about how the Go compiler works.

> For anyone who does not know, a compiler is a translator: it converts a piece of code in a language (in our case, Go) to an equivalent piece of code in another language (in our case, assembly code, whose specific instructions depend on the specific architecture we want to run the executable in). Hopefully, the meaning of what we wrote in Go (the behavior of our program), is maintained throughout the process of translation.

Reviewer 2: You are already writing the text here, so I'd perhaps avoid saying "For anyone who does not know". It might make a beginner feel bad, that they "should" know what a compiler is. Perhaps better to just write "Essentially, a compiler is .."

Author: Completely true, thanks for the suggestion!

## Rewrite rules

One of the benefits of using SSA is that it is easy to match common patterns and rewrite them with optimized versions. For example, let's say that we have the following Go code:

```go
var five = 5
var a = five * 16
```

We can compute the value of `a` by literally making the multiplication `5 times 16`. But it is known that multiplication instructions are quite CPU-expensive. What if we could optimize this line of code by avoiding the multiplication?

It turns out we can, because `16` is a power of `2`. And multiplying by `2^n` is the same as shifting the number to the left `n` places. This is easy to see in binary, where 5 is:

```
00000101
```

If we multiply 5 by 2, we get 10, which in binary is:

```
00001010
```

And again by 2, we get 20:

```
00010100
```

See how we are shifting everything to the left each time we multiply by 2? Then, multiplying by 16 is the same as shifting to the left 4 times, as 16 is `2^4`:

```
00000101 // 5
01010000 // 5*(2^4) = 80

> ```

Reviewer 2: I feel like this level of detail is unnecessary. We can just skip this part with just a one-line explanation that doubling is the same as a single left shift.

Author: I agree. I erred on the side of over-explaining, and this is a good example. Changed, thanks!

-   The new instruction, `(Lsh8x64 <t> n (Const64 <typ.UInt64> [log8(c)]))`, which shifts `n` to the left `log(c)` places.

It is ok if you do not understand every specific detail of the instructions, we will see them in detail a bit later in this post. Here, for example, we can note that the `8` suffixes make reference to the fact that this rule only matches multiplications of 8-bit variables. The other highlighted rules in the link to the `generic.rules` file cover the cases for variables of different bit widths.

For now, it's enough to understand that the rules match a specific pattern of instructions, and rewrite those instructions with a new one that is hopefully more efficient. Also, it is important to know that there are generic rules, as the one we just studied, which are applied regardless of the specific architecture we are compiling for, and architecture-specific rules, such as the ones in the [AMD64.rules file](https://github.com/golang/go/blob/master/src/cmd/compile/internal/ssa/gen/AMD64.rules), which take advantage of the specifics of each architecture, further optimizing the code.

## Connecting all together

> In order to investigate the issue and solve it, we will need to set up the development environment for the Go compiler. I encourage you to read the official [documentation](https://golang.org/doc/contribute): it is comprehensive and easy to understand. There are some particularities to the process, as the need to use Gerrit instead of Github, but it is really straightforward.

Reviewer 2: It is not really necessary to use Gerrit. You can just use GitHub too. https://github.com/golang/go/wiki/GerritBot#workflow

Author: That's true, there's no "need" to use it, but [it is encouraged](https://golang.org/doc/contribute#sending_a_change_gerrit) (although using Github for newcomers is [also encouraged](https://golang.org/doc/contribute#sending_a_change_github)). I left the phrase without the explanation between the commas. Thanks!
